### PR TITLE
driver: espi: it8xxx2: enable ESPI_OOB_CHANNEL

### PR DESCRIPTION
**Pull request by Dino-Li**
This enable eSPI out-of-band channel.

Signed-off-by: Dino Li <Dino.Li@ite.com.tw>

### CODE REVIEW

**Review by brockus-zephyr — APPROVED** [review]

File: drivers/espi/espi_it8xxx2.c [hunk 1]
```diff
@@ -687,22 +696,130 @@ static int espi_it8xxx2_write_lpc_request(const struct device *dev,
...
-	ARG_UNUSED(dev);
+	const struct espi_it8xxx2_config *const config = DRV_CONFIG(dev);
+	struct espi_slave_regs *const slave_reg =
+		(struct espi_slave_regs *)config->base_espi_slave;
+	struct espi_queue1_regs *const queue1_reg =
+		(struct espi_queue1_regs *)config->base_espi_queue1;
+	struct espi_oob_msg_packet *oob_pckt =
+		(struct espi_oob_msg_packet *)pckt->buf;
+	uint16_t oob_msg_len = oob_pckt->len_bit_7_0 |
+				((oob_pckt->tag_len_bit_11_8 & 0xf) << 8);
```
Comment (keith-zephyr) [inline drivers/espi/espi_it8xxx2.c:722]: You should also verify that `pckt->len == oob_msg_len + 3`.

File: drivers/espi/espi_it8xxx2.c [hunk 2]
```diff
@@ -1208,8 +1332,8 @@ static struct espi_it8xxx2_data espi_it8xxx2_data_0;
 static const struct espi_it8xxx2_config espi_it8xxx2_config_0 = {
 	.base_espi_slave = DT_INST_REG_ADDR_BY_IDX(0, 0),
 	.base_espi_vw = DT_INST_REG_ADDR_BY_IDX(0, 1),
-	.base_espi_queue1 = DT_INST_REG_ADDR_BY_IDX(0, 2),
-	.base_espi_queue0 = DT_INST_REG_ADDR_BY_IDX(0, 3),
+	.base_espi_queue1 = DT_INST_REG_ADDR_BY_IDX(0, 3),
+	.base_espi_queue0 = DT_INST_REG_ADDR_BY_IDX(0, 2),
```
Comment (keith-zephyr) [inline drivers/espi/espi_it8xxx2.c:1336]: Change this so the index numbers are in order (queue0 first, then queue1)

File: soc/riscv/riscv-ite/common/check_regs.c [hunk 3]
```diff
@@ -55,6 +55,15 @@ IT8XXX2_REG_OFFSET_CHECK(espi_vw_regs, VW_INDEX, 0x00);
...
 
+/* eSPI Queue 1 registers structure check */
+IT8XXX2_REG_SIZE_CHECK(espi_queue1_regs, 0xc0);
+IT8XXX2_REG_OFFSET_CHECK(espi_queue1_regs, UPSTREAM_DATA, 0x00);
+IT8XXX2_REG_OFFSET_CHECK(espi_queue1_regs, PUT_FLASH_NP_DATA, 0x80);
+
+/* eSPI Queue 0 registers structure check */
+IT8XXX2_REG_SIZE_CHECK(espi_queue0_regs, 0xd0);
+IT8XXX2_REG_OFFSET_CHECK(espi_queue0_regs, PUT_OOB_DATA, 0x80);
+
```
Comment (keith-zephyr) [inline soc/riscv/riscv-ite/common/check_regs.c:66]: Reorder to put queue0 first.

File: soc/riscv/riscv-ite/common/chip_chipregs.h [hunk 4]
```diff
@@ -2291,6 +2291,32 @@ struct espi_vw_regs {
 	/* 0x98-0x99: Reserved3 */
 	volatile uint8_t reserved3[2];
 };
+
+/*
+ * eSPI Queue 1 registers
+ */
```
Comment (keith-zephyr) [inline soc/riscv/riscv-ite/common/chip_chipregs.h:2297]: Reorder to put queue0 first.

File: drivers/espi/espi_it8xxx2.c [hunk 5]
```diff
@@ -687,22 +696,130 @@ static int espi_it8xxx2_write_lpc_request(const struct device *dev,
...
+		oob_pckt->data_byte[i] = queue0_reg->PUT_OOB_DATA[i];
+	}
+
+	return 0;
+}
+
+static void espi_it8xxx2_oob_init(const struct device *dev)
+{
+	const struct espi_it8xxx2_config *const config = DRV_CONFIG(dev);
+	struct espi_it8xxx2_data *const data = DRV_DATA(dev);
```
Comment (keith-zephyr) [inline drivers/espi/espi_it8xxx2.c:812]: As a follow up PR, please eliminate the DRV_CONFIG() and DRV_DATA() macros in the entire file.

File: drivers/espi/espi_it8xxx2.c [hunk 6]
```diff
@@ -687,22 +696,130 @@ static int espi_it8xxx2_write_lpc_request(const struct device *dev,
...
+/* eSPI cycle type field */
+#define ESPI_OOB_CYCLE_TYPE          0x21
+#define ESPI_OOB_TAG                 0x00
+#define ESPI_OOB_TIMEOUT_MS          200
+
+struct espi_oob_msg_packet {
+	uint8_t cycle_type;
+	uint8_t tag_len_bit_11_8;
+	uint8_t len_bit_7_0;
+	uint8_t data_byte[ESPI_IT8XXX2_OOB_MAX_PAYLOAD_SIZE];
```
Comment (keith-zephyr) [inline drivers/espi/espi_it8xxx2.c:708]: You don't need to allocate the payload here.  Use `uint8_t data_byte[0];` to create the data_byte field at the correct offset, but without allocating data.

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

File: drivers/espi/espi_it8xxx2.c [hunk 7]
```diff
@@ -687,22 +696,130 @@ static int espi_it8xxx2_write_lpc_request(const struct device *dev,
...
 {
-	/* TODO: implement me... */
-	ARG_UNUSED(dev);
+	const struct espi_it8xxx2_config *const config = DRV_CONFIG(dev);
+	struct espi_slave_regs *const slave_reg =
+		(struct espi_slave_regs *)config->base_espi_slave;
+	struct espi_queue1_regs *const queue1_reg =
+		(struct espi_queue1_regs *)config->base_espi_queue1;
+	struct espi_oob_msg_packet *oob_pckt =
+		(struct espi_oob_msg_packet *)pckt->buf;
```
[the remark this answers is not in the document]
Comment (Dino-Li) [reply drivers/espi/espi_it8xxx2.c:723]: Done.

File: drivers/espi/espi_it8xxx2.c (hunk 1 above)
Comment (Dino-Li) [reply drivers/espi/espi_it8xxx2.c]: Done.

File: drivers/espi/espi_it8xxx2.c (hunk 2 above)
Comment (Dino-Li) [reply drivers/espi/espi_it8xxx2.c]: Done.

File: soc/riscv/riscv-ite/common/check_regs.c (hunk 3 above)
Comment (Dino-Li) [reply soc/riscv/riscv-ite/common/check_regs.c]: Done.

File: soc/riscv/riscv-ite/common/chip_chipregs.h (hunk 4 above)
Comment (Dino-Li) [reply soc/riscv/riscv-ite/common/chip_chipregs.h]: Done.

File: drivers/espi/espi_it8xxx2.c (hunk 5 above)
Comment (Dino-Li) [reply drivers/espi/espi_it8xxx2.c:827]: Ack.

File: drivers/espi/espi_it8xxx2.c (hunk 6 above)
Comment (Dino-Li) [reply drivers/espi/espi_it8xxx2.c]: Done.

**Comment by Dino-Li** [comment]
@keith-zephyr - ping...

**Review by keith-zephyr — APPROVED** [review]

**Pull request merged**